File: Week 03/id_011/search-in-rotated-sorted-array.go
package algorithm00401

func search(nums []int, target int) int {
	l, r := 0, len(nums)
	for l < r {
		mid := l + (r-l)/2
		if nums[mid] == target {
			return mid
		} else if nums[l] == target {
			return l
		} else if nums[r-1] == target {
			return r - 1
		}
		if nums[l] < nums[mid] {
			if target > nums[mid] || nums[l] > target {
				l = mid + 1
			} else {
				r = mid
			}
		} else {
			if nums[mid] > target || target > nums[r-1] {
				r = mid
			} else {
				l = mid + 1
			}
		}
	}
	return -1
}

File: Week 03/id_081/SearchInRotatedSortedArray.java
/**
 * SearchInRotatedSortedArray
 */
public class SearchInRotatedSortedArray {

    public int search(int[] nums, int target) {
        int lo = 0;
        int hi = nums.length - 1;

        while (lo < hi) {
            int mid = (lo + hi) / 2;
            // 当[0,mid]有序时,向后规约条件
            if (nums[0] <= nums[mid] && (target > nums[mid] || target < nums[0])) {
                lo = mid + 1;
                // 当[0,mid]发生旋转时，向后规约条件
            } else if (nums[0] > nums[mid] && target > nums[mid] && target < nums[0]) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo == hi && nums[lo] == target ? lo : -1;
    }
}